cranelift/codegen/src/isa/x64/lower.isle

```diff
@@ -0,0 +1,775 @@
+
+;;;; Rules for `iconst` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; `i128`
+(rule (lower (with_type $I128
+                        (iconst (u64_from_imm64 x))))
+      (value_regs (imm $I64 x)
+                  (imm $I64 0)))
+
+;; `i64` and smaller.
+(rule (lower (with_type (fits_in_64 ty)
+                        (iconst (u64_from_imm64 x))))
+      (value_reg (imm ty x)))
+
+;;;; Rules for `bconst` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; `b128`
+
+(rule (lower (with_type $B128
+                        (bconst $false)))
+      (value_regs (imm $B64 0)
+                  (imm $B64 0)))
+
+(rule (lower (with_type $B128
+                        (bconst $true)))
+      (value_regs (imm $B64 1)
+                  (imm $B64 0)))
+
+;; `b64` and smaller.
+
+(rule (lower (with_type (fits_in_64 ty)
+                        (bconst $false)))
+      (value_reg (imm ty 0)))
+
+(rule (lower (with_type (fits_in_64 ty)
+                        (bconst $true)))
+      (value_reg (imm ty 1)))
+
+;;;; Rules for `null` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(rule (lower (with_type ty (null)))
+      (value_reg (imm ty 0)))
+
+;;;; Rules for `iadd` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; SSE.
+
+(rule (lower (with_type (multi_lane 8 16)
+                        (iadd x y)))
+      (value_reg (paddb (put_in_reg x)
+                        (put_in_reg_mem y))))
+
+(rule (lower (with_type (multi_lane 16 8)
+                        (iadd x y)))
+      (value_reg (paddw (put_in_reg x)
+                        (put_in_reg_mem y))))
+
+(rule (lower (with_type (multi_lane 32 4)
+                        (iadd x y)))
+      (value_reg (paddd (put_in_reg x)
+                        (put_in_reg_mem y))))
+
+(rule (lower (with_type (multi_lane 64 2)
+                        (iadd x y)))
+      (value_reg (paddq (put_in_reg x)
+                        (put_in_reg_mem y))))
+
+;; `i128`
+(rule (lower (with_type $I128 (iadd x y)))
+      ;; Get the high/low registers for `x`.
+      (let ((x_regs ValueRegs (put_in_regs x))
+            (x_lo Reg (value_regs_get x_regs 0))
+            (x_hi Reg (value_regs_get x_regs 1)))
+        ;; Get the high/low registers for `y`.
+        (let ((y_regs ValueRegs (put_in_regs y))
+              (y_lo Reg (value_regs_get y_regs 0))
+              (y_hi Reg (value_regs_get y_regs 1)))
+          ;; Do an add followed by an add-with-carry.
+          (with_flags (add_with_flags $I64 x_lo (RegMemImm.Reg y_lo))
+                      (adc $I64 x_hi (RegMemImm.Reg y_hi))))))
+
+;; Add two registers.
+(rule (lower (with_type (fits_in_64 ty)
+                        (iadd x y)))
+      (value_reg (add ty
+                      (put_in_reg x)
+                      (RegMemImm.Reg (put_in_reg y)))))
+
+;; Add a register and an immediate.
+
+(rule (lower (with_type (fits_in_64 ty)
+                        (iadd x (simm32_from_value <ty y))))
```

Comment:
This is an "in-arg" to the left-hand-side extractor; the basic idea is that while the arg-positions of left-hand-side patterns are mostly "outputs" of whatever the matcher is doing (the inverse of a function application on the RHS), some matcher implementations will want an *input* (in addition to the thing they're matching on) to configure how they match. So here I think `simm32_from_value` is a pattern-matcher that is parameterized on a type.

(If one were doing this in e.g. Prolog it'd be just another argument, and dataflow could flow either way during unification. Here we have "argument polarity" instead, which one can think of as an approximation of this...)

I did mostly intend this to be a low-level feature that we could use to define other sorts of extractors/matchers... 

@fitzgen: I wonder if there's a way to better write this? It looks like the `ty` in-arg eventually gets down to `to_simm32` [here](https://github.com/fitzgen/wasmtime/blob/435df0f9d2191d050e29a93a01df1517e4edfceb/cranelift/codegen/src/isa/x64/lower/isle.rs#L413) which just checks if `ty` is less than 64 bits, and skips an "is sign extendable" check if so; but it looks like this may actually be unnecessary (I don't fully remember the semantics of simm32 values on 32 vs 64 bit instructions though)? Or could we just split out the 8/16/32 and 64 bit cases (one of the advantages of just having a list of patterns even!)?

Reply:
Alternately, we propagate arg polarity to the parser and parse in-args as expressions without the `<`, but then that makes this a type-dependent parse (a la C++) which is kind of not great. We could do it though if we require decls to come before uses of terms...

Reply:
> Or could we just split out the 8/16/32 and 64 bit cases (one of the advantages of just having a list of patterns even!)?

What's nice about having polymorphic rules instead of a separate rule for each of `i{8,16,32,64}` is that they actually end up as a single branch in the trie/generated code, instead of four different branches.

> I wonder if there's a way to better write this? It looks like the `ty` in-arg eventually gets down to `to_simm32` [here](https://github.com/fitzgen/wasmtime/blob/435df0f9d2191d050e29a93a01df1517e4edfceb/cranelift/codegen/src/isa/x64/lower/isle.rs#L413) which just checks if `ty` is less than 64 bits, and skips an "is sign extendable" check if so; but it looks like this may actually be unnecessary

Yeah I think we could just always do the "is sign extendable" check and it would be equivalent.

Reply:
Ah, I see. Fwiw, I'd find `=ty` more intuitive syntax for this than `<ty`.

Reply:
`=variable` is already used for unifying two variables, eg

```lisp
(rule (has_type ty (ixor x =x))
    (imm ty 0))
```

Reply:
@fitzgen since we're hoping it will mostly be a lower-level thing that we build on top of, we could make the syntax more explicit, maybe: what do you think about something like `(eval ...)` or `(matcher-input ...)`?

(Basically, I too would be happy to see `<ty` die; it was literally just the first punctuation character I thought of and I wish I had chosen something better initially!)

Reply:
`simm32` immediates no longer need input parameters, fwiw. They were the only use of input args to extractors in all of the lowering code.

One thing I was thinking, and maybe this is wild, was to always have extractors use `[...]` and always have expressions use `(...)` but this doesn't really solve the case where a simple, non-operation expression is an input.

Or maybe, since we aren't using input polarity arguments for extractors anymore, we simply don't need them and can remove them from the language?

Reply:
It may be different in terms of what's actually going on in the DSL, however as someone just reading the code, this `=x` and the `<ty` above kinda feel like they're doing the same conceptual thing. Whether it's a type or a variable, the prefix means "don't bind a new thing here, reuse an existing binding". So I wonder if it would work to use prefix-`=` syntax for both.

Reply:
Wouldn't there be a way to have very explicit syntax with another sub-expression or an annotation, instead of using an operator which significance has to be looked up in documentation and learned about? Ad-hoc syntax isn't self-explanatory, and as a matter of fact it is spurious extra complexity for newcomers, especially if it _can_ be avoided by using another more explicit syntax.

Reply:
Agreed with @bnjbvr; I'd like to re-boost my suggestion above for `(matcher-input ...)` if we end up needing this feature still. I'm also hoping that, if it remains, we can keep its use confined to just whatever "standard library helpers" need it, if that, and keep it out of ordinary patterns entirely.

Also agreed with @fitzgen that given that we found a way to avoid it here, maybe we can remove it; it nudges the language more toward "general-purpose constraint solver" (ie the Prolog direction) and less toward "patterns of instructions", when the latter is really the simplicity and approachability we're going for, so with whatever influence I still have on the language design, I'd like to push to undo my mistake :-)

Reply:
For sure. I'm not going to deal with it right now, in this PR, but if we don't end up removing this feature from the DSL completely, we can use `(eval ...)` or `(matcher-input ...)` or something.